win_perf_counters/kernel32.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2010 The win Authors. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+// 1. Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+// 2. Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+// 3. The names of the authors may not be used to endorse or promote products
+//    derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND ANY EXPRESS OR
+// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+// OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+// IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+// NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+// THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// This is the official list of 'win' authors for copyright purposes.
+//
+// Alexander Neumann <an2048@googlemail.com>
+// Joseph Watson <jtwatson@linux-consulting.us>
+// Kevin Pors <krpors@gmail.com>
+
+// +build windows
+
+package win_perf_counters
```

Comment:
Go package names should not contain underscores.
See. https://blog.golang.org/package-names

I would suggest to name the package `performancecounters`. Since I'm working on a implementation to retrieve CPU stats and will probably use the same approach as you did, I could extend that structure.

```
performancecounters/mem.go
performancecounters/mem_windows.go
performancecounters/kernel32.go
...
```

I can then add my `cpu.go`  and `cpu_windows.go`

Reply:
Yes, I know about the package name style, it was just copied from the telegraf.
Renamed it. 
But I don't get what is the reason to put `mem.go`, `mem_windows.go` and `cpu*` inside the `perfcounters` folder? It

Reply:
> Yes, I know about the package name style, it was just copied from the telegraf.

Glad we started with PRs then. If you know about the convention I wonder why you don't apply it then.

> But I don't get what is the reason to put mem.go, mem_windows.go and cpu* inside the perfcounters folder?

We group them in a package because their functionality is related. Retrieving CPU and memory info. Technically only windows implementation uses "real" performance counters but I think the package name still fits. Does that make sense or what you struggle to understand?

Reply:
1. Both `mem.go` and `cpu.go` includes a lot of Unix stuff 
2. If you move `mem_windows.go` and `cpu_windows.go` into `perfcounters` package directory, where do you plan to require `prefcounter` package?

Reply:
> Both mem.go and cpu.go includes a lot of Unix stuff

I know. The windows implementation is in the `_windows.go` files. But from the point of view of our application we still have to put the logic somewhere together. I'm open for another name suggestion to avoid collision with the windows performance counters feature. I think `performancecounters` still fits our use case but to avoid confusion we can up with our own naming.

> If you move mem_windows.go and cpu_windows.go into perfcounters package directory, where do you plan to require prefcounter package?

I don't fully understand. The functionality to retrieve cpu/mem information is isolated in its own package and can be included from the `cagent` package. Maybe it's not really obvious at the moment because in the current structure everything is within one package, which is not a good practice. With putting this functionality in its own package we start to move to a more modular approach.

Reply:
I just noticed that I was a step too far in my head already.
I suggested to put `mem.go` and the other files into the `performancecounters` package because I think that's a better way to organize our code.
But I don't wanna make this PR more complicated. Leave the `mem.go` and `mem_windows.go` where it is right now. I will add my new functionality to the `performancecounters` package then and at some point we can refactor the memory implementation.

Another think I noticed: Packages should be placed under the `pkg` package.
See: https://github.com/golang-standards/project-layout
`pkg/performancecounters` in our case. Sorry have overlooked that before

Reply:
> Actually, I don't remember any lib that uses the pkg directory approach :)

We're building an application and not a standalone library. I agree that you don't necessarily need a `pkg` package in a library but when building an application it's good practice to put your own code there.

Reply:
Yes, but ```Other projects will import these libraries expecting them to work, so think twice before you put something here :-)```
This looks like just a recommendation on how to `export` internal library to be used by other projects.

This is not wrong to just use the internal package naming